### PR TITLE
Composite query

### PR DESCRIPTION
**Pull request by bdemann**

### CODE REVIEW

File: examples/composite_queries/src/canister1/canister1.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+
+canister1 = Canister1(Principal.from_str('rrkah-fqaaa-aaaaa-aaaaq-cai'))
+canister2 = Canister2(Principal.from_str('ryjl3-tyaaa-aaaaa-aaaba-cai'))
+counter: int = 0
+
+
+# Composite query calling a query
+@query
+@composite
```
Comment (lastmjs) [inline examples/composite_queries/src/canister1/canister1.py:13]: Let's consider getting rid of the decorator and just making all async queries composite queries.

**Review by lastmjs — CHANGES_REQUESTED** [review]

**Comment by bdemann** [comment]
We need to make sure that after https://github.com/demergent-labs/azle/pull/845 gets merged that we update the tests to point to the azle tests.

File: examples/composite_queries/src/canister1/canister1.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+    return {
+        'ok': result.ok
+    }
+
+
+# Composite query that modifies the state. Should revert after the call is done
+@query
+def inc_counter() -> nat:
+    global counter
```
Comment (lastmjs) [inline examples/composite_queries/src/canister1/canister1.py:112]: global I believe is not necessary here

File: examples/composite_queries/src/canister1/canister1.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    counter += 1
+    return counter
+
+
+# Composite query calling queries on the same canister. SHOULDN'T WORK
+@query
+def inc_canister1() -> Async[NatQueryResult]:
+    global counter
```
Comment (lastmjs) [inline examples/composite_queries/src/canister1/canister1.py:120]: global I believe is not necessary here

File: examples/composite_queries/src/canister1/canister1.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+    return {
+        'ok': counter + canister1_a.ok + canister1_b.ok
+    }
+
+
+# Composite query calling queries that modify the state
+@query
+def inc_canister2() -> Async[NatQueryResult]:
+    global counter
```
Comment (lastmjs) [inline examples/composite_queries/src/canister1/canister1.py:143]: global I believe is not necessary here

File: examples/composite_queries/src/canister1/canister1.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    counter += 1
+
+    canister2_a: CanisterResult[nat] = yield canister2.inc_counter()
```
Comment (lastmjs) [inline examples/composite_queries/src/canister1/canister1.py:146]: If you could use the same naming conventions for these variables as in Azle

File: examples/composite_queries/src/canister2/canister2.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+from src.canister3.types import Canister3
+
+canister3 = Canister3(Principal.from_str('r7inp-6aaaa-aaaaa-aaabq-cai'))
+
+counter: nat = 0
+
+
+@query
+def inc_counter() -> nat:
+    global counter
```
Comment (lastmjs) [inline examples/composite_queries/src/canister2/canister2.py:12]: global should not be necessary here

File: examples/composite_queries/src/canister2/canister2.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+from src.canister1.types import StringQueryResult
+from src.canister3.types import Canister3
+
+canister3 = Canister3(Principal.from_str('r7inp-6aaaa-aaaaa-aaabq-cai'))
+
+counter: nat = 0
+
+
+@query
+def inc_counter() -> nat:
```
Comment (lastmjs) [inline examples/composite_queries/src/canister2/canister2.py:11]: This is a composite query in azle but is not here in kybra

File: examples/composite_queries/src/canister1/canister1.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        'ok': result.ok
+    }
+
+
+# Manual composite query calling a manual query
+@query
+def totally_manual_query() -> Async[manual[StringQueryResult]]:
+    result: CanisterResult[str] = yield canister2.manual_query()
+
+    if result.err is not None:
```
Comment (lastmjs) [inline examples/composite_queries/src/canister1/canister1.py:46]: Why does Azle look so different? In Azle we are returning without unwrapping the result...why is that?

File: examples/composite_queries/src/canister1/canister1.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        'ok': result.ok
+    }
+
+
+# Composite query being called by an update method. SHOULDN'T WORK
+@update
+def simple_update() -> Async[StringQueryResult]:
+    result: CanisterResult[str] = yield canister2.deep_query()
+
+    if result.err is not None:
```
Comment (lastmjs) [inline examples/composite_queries/src/canister1/canister1.py:99]: Why does Azle look so different with the unwrapping?

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: examples/composite_queries/src/canister1/canister1.py (hunk 2 above)
Comment (bdemann) [reply examples/composite_queries/src/canister1/canister1.py:112]: It is necessary without it pylance gives a '"counter" is unbound' error

File: examples/composite_queries/src/canister1/canister1.py (hunk 3 above)
Comment (bdemann) [reply examples/composite_queries/src/canister1/canister1.py:120]: It is necessary without it pylance gives a '"counter" is unbound' error

File: examples/composite_queries/src/canister1/canister1.py (hunk 4 above)
Comment (bdemann) [reply examples/composite_queries/src/canister1/canister1.py:143]: It is necessary without it pylance gives a '"counter" is unbound' error

File: examples/composite_queries/src/canister2/canister2.py (hunk 6 above)
Comment (bdemann) [reply examples/composite_queries/src/canister2/canister2.py:12]: It is necessary without it pylance gives a '"counter" is unbound' error

File: examples/composite_queries/src/canister2/canister2.py (hunk 7 above)
Comment (bdemann) [reply examples/composite_queries/src/canister2/canister2.py:11]: I'll need some help with this one. Pylance is giving me some weird errors, that I cannot figure out.

**Review by lastmjs — DISMISSED** [review]

**Review by lastmjs — APPROVED** [review]

**Pull request merged**